celery/app/task.py

```diff
@@ -959,13 +959,34 @@ def replace(self, sig):
             stamped_headers = self.request.stamped_headers.copy()
             stamps = self.request.stamps.copy()
             stamped_headers.extend(sig.options.get('stamped_headers', []))
+            stamped_headers = list(set(stamped_headers))
             stamps.update({
                 stamp: value
                 for stamp, value in sig.options.items() if stamp in sig.options.get('stamped_headers', [])
             })
             sig.options['stamped_headers'] = stamped_headers
             sig.options.update(stamps)
 
+            links = sig.options['link'] if 'link' in sig.options else []
+            links.extend(sig.options['link_error'] if 'link_error' in sig.options else [])
+
+            if hasattr(sig, "tasks"):
+                tasks = sig.tasks
+                if isinstance(tasks, group):
```

Comment:
What happens if `tasks` is a list or a tuple?
What happens if it is a `regen`?

We shouldn't ignore such cases.

Reply:
Your new test cases should cover this. Let's see if the PR tests pass with the new edge cases.

Reply:
But it should also work as the canvas primitive already handles the `tuple` and `regen` cases, but the added matrix for the test will cover this 100%

Reply:
@thedrow Yes, the new test cases are passing with this code which confirms your original question is resolved.
> What happens if tasks is a list or a tuple?

It works.

> What happens if it is a regen?

It works.

> We shouldn't ignore such cases.

We do not :)